Python: add JSONSchema encoding tutorial (#463)
foxglove/mcap: python/docs/jsonschema_tutorial.rst

------------

In this tutorial we'll take a publicly available dataset and demonstrate how to convert it from a
CSV format to an MCAP file. We'll use Python and `JSON Schema <https://json-schema.org/>`_ to
quickly create messages without requiring extra serialization libraries or generated code.

You can run all the code in this tutorial or reference it
`in the MCAP repo <https://github.com/foxglove/mcap/tree/main/python/examples/jsonschema/pointcloud_csv_to_mcap.py>`_.

Decoding the data source
------------------------

This tutorial uses the public CSV dataset
`"Sydney Urban Objects Dataset" <https://www.acfr.usyd.edu.au/papers/SydneyUrbanObjectsDataset.shtml>`_,
released by the **Australian Centre for Field Robotics** at the University of Sydney.
This is a collection of CSV files containing point clouds collected from objects on the street.

Decoding the data is pretty simple, thanks to Python's built-in ``csv`` and ``datetime`` libraries:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-csv-decode-start
    :end-before: # tutorial-csv-decode-end

Creating a ``foxglove.PointCloud``
----------------------------------

> In order to view this point cloud, we need encode these points in a way that Foxglove Studio

Reviewer: ```suggestion
To view this point cloud, we need to encode these points in a way that Foxglove Studio
```

> understands. To facilitate this, Foxglove Studio provides a collection of schemas for messages that

Reviewer: ```suggestion
understands. To facilitate this, Foxglove Studio provides a collection of message schemas that
```

JSON ``PointCloud`` instance, using the provided 
`schema <https://github.com/foxglove/schemas/blob/main/schemas/jsonschema/PointCloud.json>`_.

Let's start with encoding the point data. The schema expects a single ``base64``-encoded buffer
containing all point data, and some metadata describing how to decode it.

The CSV data contains one timestamp and four floating-point data fields per point.
``foxglove.PointCloud`` uses one timestamp for the whole point cloud, so we'll use the first
point's timestamp. We'll pack each field as a four byte single-precision little-endian float.

We start by describing our data layout in the ``foxglove.PointCloud`` message:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-point-layout-start
    :end-before: # tutorial-point-layout-end
    :dedent:

And pack the points using Python's built-in ``struct`` and ``base64`` libraries.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-pack-points-start
    :end-before: # tutorial-pack-points-end
    :dedent:

In Foxglove Studio, each 3D object exists in its own coordinate frame. A point cloud's ``frame_id``

> identifies which coordinate frame it belongs in, and its ``pose`` determines it's relative position

Reviewer: ```suggestion
identifies which coordinate frame it belongs in, and its ``pose`` determines its relative position
```

> from the center of that coordinate frame.

Reviewer: ```suggestion
from that coordinate frame's center.
```

Since we will only have one coordinate frame in our MCAP file, we can choose any string as our

> ``frame_id``, and use the identity pose to place our point cloud in the center of it.

Reviewer: ```suggestion
``frame_id``, and use the identity pose to place our point cloud in its center.
```

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-pose-frame-id-start
    :end-before: # tutorial-pose-frame-id-end
    :dedent:

We'll leave the ``timestamp`` field for later, when we write the messages into the MCAP file.

Writing the MCAP file
---------------------

Now that the point cloud is built, we can write it into an MCAP file. We'll start with some imports
from the `Python MCAP library <https://github.com/foxglove/mcap/tree/main/python>`_:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-mcap-imports-start
    :end-before: # tutorial-mcap-imports-end
    :dedent:

Let's open a file where we'll write our output MCAP data. First, we'll need to write our header.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-header-start
    :end-before: # tutorial-write-header-end
    :dedent:

.. note::
  We can choose whatever name we want for ``library``. We're not using one of the
  `MCAP well-known profiles <https://github.com/foxglove/mcap/blob/main/docs/specification/appendix.md#well-known-profiles>`_,

>   so we use our own custom profile name. Custom profile names must begin with ``x-``.

Reviewer 2: See https://github.com/foxglove/mcap/issues/370...

Next, create a "channel" of messages to contain our point cloud. The schema
name and content informs Foxglove Studio that it can parse and display this message as a point
cloud.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-channel-start
    :end-before: # tutorial-write-channel-end
    :dedent:

Next, we write our messages. If we only wrote one message, our MCAP file would have an infinitely

> short duration. To address that, let's write our point cloud message a few times with successive

Author: the single-message MCAP wouldn't actually display anything in Studio when I tried it. I'm not sure if we consider this a bug, but it's a pretty specific use-case.

> short duration. To address that, let's write our point cloud message a few times with successive

Reviewer 2: Sounds like a bug. Is it the same as https://github.com/foxglove/studio/pull/3828? Guess not because that's ros2 only. Well it's similar at least 😄

********************************
Tutorial: Converting CSV to MCAP
********************************

Introduction
------------

> In this tutorial we'll take a publicly-available dataset and demonstrate a simple case

Reviewer: ```suggestion
In this tutorial, we'll take a publicly available dataset and demonstrate how to convert CSV data...
```

Again, would recommend mentioning JSON Schema in here somewhere.

of converting CSV data to MCAP for viewing in Foxglove Studio. All of the code from this tutorial
is runnable, and can be found
`in the MCAP repo <https://github.com/foxglove/mcap/tree/main/python/examples/jsonschema/pointcloud_csv_to_mcap.py>`_.

Decoding the Data Source
------------------------

This tutorial uses the public CSV dataset "Sydney Urban Objects Dataset",
released by the Australian Centre for Field Robotics at the University of Sydney, NSW, Australia.
The original dataset can be downloaded from
`their website <https://www.acfr.usyd.edu.au/papers/SydneyUrbanObjectsDataset.shtml>`_. This is a
collection of CSV files containing pointclouds collected from objects on the street.
Decoding the data is pretty simple, thanks to Python's built-in ``csv`` and ``datetime`` libraries:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-csv-decode-start
    :end-before: # tutorial-csv-decode-end

> Creating a ``foxglove.Pointcloud``

Reviewer: ```suggestion
Creating a ``foxglove.PointCloud``
```

Reviewer: Link to JSON schema in the foxglove/schemas repo

``foxglove.PointCloud`` uses one timestamp for the whole pointcloud, so we'll just use the first
point for that. We'll pack each field as a four byte single-precision little-endian float.

We start by describing our data layout in the ``foxglove.PointCloud`` message:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-point-layout-start
    :end-before: # tutorial-point-layout-end

And pack the points using the Python built-in ``struct`` and ``base64`` libraries.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-pack-points-start
    :end-before: # tutorial-pack-points-end

To set the pointcloud timestamp, the ``foxglove.PointCloud`` schema expects a pair of seconds
and nanoseconds:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-timestamp-start
    :end-before: # tutorial-timestamp-end

And finally we set an arbitrary pose and ``frame_id``.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-pose-frame-id-start
    :end-before: # tutorial-pose-frame-id-end

> Writing the MCAP

Reviewer: This post goes back and forth between calling it "the MCAP"/ "an MCAP" vs. "the MCAP file" / "an MCAP file". I think the latter makes more sense.

----------------

> Now that the pointcloud is built, we can write it into an MCAP. We'll start with some imports from

Reviewer: "point cloud" should be two words.

> the python MCAP library:

Reviewer: ```suggestion
the Python MCAP library:
```

Add link as well.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-mcap-imports-start
    :end-before: # tutorial-mcap-imports-end

Then we'll open a file for writing and write a header. Note that we can chose whatever name we
want for both ``profile`` and ``library``, but the profile must start with ``x-``.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-header-start
    :end-before: # tutorial-write-header-end

> After writing a header, we can create a "channel" of messages to contain our pointcloud. The schema

Reviewer: ```suggestion
Next, create a "channel" of messages to contain our point cloud. The schema
```

name and content informs Foxglove Studio that it can parse and display this message as a pointcloud.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-channel-start
    :end-before: # tutorial-write-channel-end

Finally, we can write our message and ``finish()`` the MCAP writer. ``finish()`` writes the summary
and footer to the MCAP file.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-message-start
    :end-before: # tutorial-write-message-end

> That's it! Once this is done, we have a valid MCAP file containing a pointcloud message. We can

Reviewer: ```suggestion
That's it! We now have a valid MCAP file containing a point cloud message. We can
```

> check this with the MCAP CLI tool:

Reviewer: Link to CLI tool page: https://github.com/foxglove/mcap/tree/main/go/cli/mcap

.. code-block:: bash

    $ mcap info output.mcap 
    library: my-excellent-library
    profile: x-jsonschema
    messages: 1
    duration: 0s
    start: 2011-11-04T01:32:10.881030912+11:00 (1320330730.881030912)
    end: 2011-11-04T01:32:10.881030912+11:00 (1320330730.881030912)
    compression:
            zstd: [1/1 chunks] (40.38%) 
    channels:
            (1) /pointcloud  1 msgs (+Inf Hz)   : foxglove.PointCloud [jsonschema]  
    attachments: 0
    $ mcap doctor output.mcap
    Examining output.mcap

> We can view the result in Foxglove Studio by navigating to `<https://studio.foxglove.dev>`_

Reviewer: ```suggestion
View your point cloud in Foxglove Studio by navigating to `<https://studio.foxglove.dev>`_
```

Reviewer: Must mention adding a 3D panel, as that visualization will not automatically add itself.

*************************************************
Tutorial: Converting CSV to MCAP with JSON Schema
*************************************************

Introduction
------------

> In this tutorial we'll take a publicly available dataset and demonstrate how to convert CSV

Reviewer: ```suggestion
In this tutorial, we'll take a publicly available dataset and demonstrate how to convert it from a CSV format to an MCAP file
```

data to MCAP. We'll use Python and `JSON Schema <https://json-schema.org/>`_ to get up and running

> quickly. All of the code from this tutorial is runnable, and can be found

Reviewer: ```suggestion
quickly. You can run all the code in this tutorial, or reference it
```

`in the MCAP repo <https://github.com/foxglove/mcap/tree/main/python/examples/jsonschema/pointcloud_csv_to_mcap.py>`_.

> Decoding the Data Source

Reviewer: ```suggestion
Decoding the data source
```

Reviewer: Using sentence case to be consistent (you later use "MCAP file" in a header).

------------------------

> This tutorial uses the public CSV dataset "Sydney Urban Objects Dataset",

Reviewer: ```suggestion
This tutorial uses the public CSV dataset `"Sydney Urban Objects Dataset" <https://www.acfr.usyd.edu.au/papers/SydneyUrbanObjectsDataset.shtml>`_,
```

> released by the **Australian Centre for Field Robotics** at the University of Sydney, NSW, Australia.

Reviewer: ```suggestion
released by the **Australian Centre for Field Robotics** at the University of Sydney.
```

> The original dataset can be downloaded from

Reviewer: Remove this sentence.

`their website <https://www.acfr.usyd.edu.au/papers/SydneyUrbanObjectsDataset.shtml>`_. This is a

> collection of CSV files containing point clouds collected from objects on the street.

Reviewer: "The dataset contains point cloud data collected on the road [by autonomous cars?]."

Decoding the data is pretty simple, thanks to Python's built-in ``csv`` and ``datetime`` libraries:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-csv-decode-start
    :end-before: # tutorial-csv-decode-end

Creating a ``foxglove.PointCloud``
----------------------------------

In order to view this point cloud, we need encode these points in a way that Foxglove Studio
understands. To facilitate this, Foxglove Studio provides a collection of schemas for messages that
it knows how to display. These are available for a variety of serializations
`on GitHub <https://github.com/foxglove/schemas>`_. For this tutorial, we'll focus on building a
JSON ``PointCloud`` instance, using the provided 
`schema <https://github.com/foxglove/schemas/blob/main/schemas/jsonschema/PointCloud.json>`_.

> Lets start with encoding the point data. The schema expects a single ``base64``-encoded buffer

Reviewer: ```suggestion
Let's start with encoding the point data. The schema expects a single ``base64``-encoded buffer
```

        "type": "object",
        "properties": {
          "name": { "type": "string", "description": "Name of the field" },
          "offset": { "type": "integer", "minimum": 0, "description": "Byte offset from start of data buffer" },
          "type": {
            "title": "NumericType: Numeric type",
            "description": "Type of data in the field. Integers are stored using little-endian byte order.",
            "oneOf": [
              { "title": "UNKNOWN", "const": 0 },
              { "title": "UINT8", "const": 1 },
              { "title": "INT8", "const": 2 },
              { "title": "UINT16", "const": 3 },
              { "title": "INT16", "const": 4 },
              { "title": "UINT32", "const": 5 },
              { "title": "INT32", "const": 6 },
              { "title": "FLOAT32", "const": 7 },
              { "title": "FLOAT64", "const": 8 }
            ]
          }
        }
      },
      "description": "Fields in the `data`"
    },
    "data": {
      "type": "string",
      "contentEncoding": "base64",
      "description": "Point data, interpreted using `fields`"
    }

> From the CSV data we have one timestamp and four floating-point data fields per point.

Reviewer: ```suggestion
The CSV data contains one timestamp and four floating-point data fields per point.
```

> ``foxglove.PointCloud`` uses one timestamp for the whole point cloud, so we'll just use the first

Reviewer: ```suggestion
``foxglove.PointCloud`` uses one timestamp for the whole point cloud, so we'll use the first
```

> point for that. We'll pack each field as a four byte single-precision little-endian float.

Reviewer: ```suggestion
point's timestamp. We'll pack each field as a four-byte, single-precision, little-endian float.
```

We start by describing our data layout in the ``foxglove.PointCloud`` message:

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-point-layout-start
    :end-before: # tutorial-point-layout-end
    :dedent:

> And pack the points using the Python built-in ``struct`` and ``base64`` libraries.

Reviewer: ```suggestion
We then pack the points using Python's built-in ``struct`` and ``base64`` libraries.
```

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-pack-points-start
    :end-before: # tutorial-pack-points-end
    :dedent:

> We set an identity pose to place our point cloud at the center of the scene, and add an arbitrary

Reviewer: Explanation for this field id?

    :start-after: # tutorial-mcap-imports-start
    :end-before: # tutorial-mcap-imports-end
    :dedent:

Let's open a file where we'll write our output MCAP data. First, we'll need to write our header.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-header-start
    :end-before: # tutorial-write-header-end
    :dedent:

.. note::
  We can choose whatever name we want for ``library``. We're not using one of the
  `MCAP well-known profiles <https://github.com/foxglove/mcap/blob/main/docs/specification/appendix.md#well-known-profiles>`_,
  so we use our own custom profile name. Custom profile names must begin with ``x-``.

Next, create a "channel" of messages to contain our point cloud. The schema
name and content informs Foxglove Studio that it can parse and display this message as a point
cloud.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-channel-start
    :end-before: # tutorial-write-channel-end
    :dedent:

Next, we write our messages. If we only wrote one message, our MCAP file would have an infinitely

> short duration. To address that, we write our point cloud message a few times with increasing

Reviewer: ```suggestion
short duration. To address that, let's write our point cloud message a few times with successive
```

timestamps.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-write-message-start
    :end-before: # tutorial-write-message-end
    :dedent:

> Finally we can invoke ``finish()`` on the MCAP writer. This will include the summary and footer

Reviewer: ```suggestion
Finally, we invoke ``finish()`` on the MCAP writer to include the summary and footer
```

in the output file.

.. literalinclude:: ../examples/jsonschema/pointcloud_csv_to_mcap.py
    :start-after: # tutorial-finish-start
    :end-before: # tutorial-finish-end
    :dedent:

That's it! We now have a valid MCAP file with 10 point cloud messages. We can
check this with the `MCAP CLI tool <https://github.com/foxglove/mcap/tree/main/go/cli/mcap>`_:

.. code-block:: bash

>     $ brew install mcap

Reviewer: Maybe break up the install and the info commands?

i.e. "Install the MCAP CLI tool using the following command... then, do xyz".

    ...
    $ mcap info output.mcap
    library: my-excellent-library
    profile: x-jsonschema
    messages: 10
    duration: 900ms
    start: 2011-11-04T01:32:10.881030912+11:00 (1320330730.881030912)
    end: 2011-11-04T01:32:11.781030912+11:00 (1320330731.781030912)
    compression:
            zstd: [1/1 chunks] (93.43%)
    channels:
            (1) /pointcloud  10 msgs (11.11 Hz)   : foxglove.PointCloud [jsonschema]
    attachments: 0
    $ mcap doctor output.mcap
    Examining output.mcap

View your point cloud in `Foxglove Studio <https://studio.foxglove.dev>`_ by

> dragging the MCAP file into the window. Add a 3D Panel and enable the **/pointcloud** topic to

Reviewer: Add link to foxglove.dev/docs/studio/panels/3d
